api/errors.go

```diff
@@ -10,39 +11,96 @@ import (
 // The canonical errors from the EigenDA gRPC API endpoints.
 //
 // Notes:
-// - We start with a small (but sufficient) subset of grpc's error codes,
-//   and expand when there is an important failure case to separate out. See:
-//   https://grpc.io/docs/guides/status-codes/
-// - Make sure that internally propagated errors are eventually wrapped in one of the
-//   user-facing errors defined here, since grpc otherwise returns an UNKNOWN error code,
-//   which is harder to debug and understand for users.
-// - See https://github.com/googleapis/googleapis/blob/ba8ea80f25d19bde8501cd51f314391f8d39bde8/google/rpc/code.proto
-//   for the mapping of grpc error codes to HTTP status codes.
-
+//   - We start with a small (but sufficient) subset of grpc's error codes,
+//     and expand when there is an important failure case to separate out. See:
+//     https://grpc.io/docs/guides/status-codes/
+//   - Make sure that internally propagated errors are eventually wrapped in one of the
+//     user-facing errors defined here, since grpc otherwise returns an UNKNOWN error code,
+//     which is harder to debug and understand for users.
+//   - See https://github.com/googleapis/googleapis/blob/ba8ea80f25d19bde8501cd51f314391f8d39bde8/google/rpc/code.proto
+//     for the mapping of grpc error codes to HTTP status codes.
 func newErrorGRPC(code codes.Code, msg string) error {
 	return status.Error(code, msg)
 }
 
+// WrapAsGRPCError wraps an error with a gRPC code and message. If the error already has a gRPC code,
+// it retains the code but combines the messages. Otherwise, it uses the provided code and combined message.
+func WrapAsGRPCError(err error, code codes.Code, msg string) error {
```

Comment:
When should this happen? Feels weird to me to have to wrap a GRPC error in another one, because a grpc status code wrapping should be terminal and returned to the user. Otherwise don't we need to define the overwrite/precedence logic for which error to return to user (eg 400 comes before 500 or not?)

Reply:
Also should prob make private so force people to use the other specific ones you wrote for each error code?

Reply:
When a server uses other services as backend, this can happen.

If the server calls `HandleFooWithServiceX(params) error`, and an error returned, this error by default should be INTERNAL error, but it's also possible the `params` are no valid which results in an error (the caller may not have sufficient information to validate `params` before calling).

In such case you may want to `WrapAsGRPCError(err, INTERNAL, "....")`: if `HandleFooWithServiceX` has an opinion about the error type, then we should use that; otherwise, it should be INTERNAL.

Reply:
I see! Thanks for the explanation.

So right, if an external service returns a 400, that should be, from our server's client's pov, converted to a 500 (we are doing something wrong).

So I feel like something simpler would be:
- rename this function something clearer (`convertExternalServiceError`?) and your explanation should be added as documentation comment)
- or just get rid of this function altogether, and force people to use the other methods (unwrap the error string from external service's error and use NewErrorInternal(err.Err()))

I personally like to keep any package's public method surface small, so would prefer to remove functions.

Reply:
it's not just external service error, it can be the validation with more info from external service failed